daemon/container_operations.go

```diff
@@ -980,6 +980,9 @@ func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName
 	if endpointConfig == nil {
 		endpointConfig = &networktypes.EndpointSettings{}
 	}
+	container.Lock()
+	defer container.Unlock()
```

Comment:
libnetwork calls CreateEndpoint/Join also doesn't have to be under the container lock. But the container object currently is touched in too many places making it difficult to refactor.

LGTM.